### ISSUE TITLE
[Question] How can i get the online client list

### ISSUE THREAD

**Issue by liangyuanpeng** [eclipse/mosquitto#2048]
Hi,Please forgive me for asking a simple question.  

I want to know who is using my mqtt broker , some info like `clientId`, `connectIP` .

If i missed any document,Remind me please!

Thanks! :)

**Comment by ralight** [comment]
You could watch the logs to see clients connecting, but there isn't a way for you to query which clients are connected at any given point.

**Comment by liangyuanpeng** [comment]
Got it, Thanks for your reply

**Comment by ralight** [comment]
No, not at the moment. You could do it in a plugin without too much trouble, but having to keep track of the client list yourself. You'll need the `connect` event that is in the `develop` branch though.

**Comment by tanner0101** [comment]
@ralight is a "connected clients" list something mosquitto would consider adding? I would be happy to put together a PR to enable this behind a conf option. Having a list of connected clients in `$SYS` would be really helpful for my use case. Checking the logs works but isn't ideal. I've also considered having the clients maintain a status topic but it's not super obvious (to me at least) how you would ensure this gets updated _and_ cleared if the client disconnects uncleanly. Doing it in the broker seems like the best option.

**Comment by ralight** [comment]
I'm opposed to it because I'm thinking about the case where 1 million clients are connected.

You can do it at the application level using a Will message to set/clear a status when the client disconnects uncleanly.

**Comment by tanner0101** [comment]
> I'm opposed to it because I'm thinking about the case where 1 million clients are connected.

Right, that makes sense. In my mind, I see this flag as being kind of like `log_type all`. It has performance drawbacks but debug benefits. It should definitely be opt-in only.

> You can do it at the application level using a Will message to set/clear a status when the client disconnects uncleanly.

That's what I was hoping I could do, too. But I can't figure out how. Hopefully I'm misunderstanding something though! My idea was to have clients post a _retained_ message to `$SYS/broker/connection/%c/state` (e.g., `1` for "connected"). When they disconnect, they would update the state to `0` and also (importantly) clear the retained message. This would require two messages if I understand correctly. One to set the status to `0` and then an empty message to clear the retained one. AFAIK you cannot update a topic and clear retained in the same message?

> You could do it in a plugin without too much trouble, but having to keep track of the client list yourself.

FWIW, I also like this option 👍

**Comment by mbenabda** [comment]
@tanner0101 no need for a retained message here. as @ralight  pointed out, you would rather do that with a will message:

In its on_connect handler, have your client send a "connected" message, followed by a "disconnected" message on the same topic with the will message flag set. 

You'd have to ensure that these messages:
- are sent with at least QoS 1
- are timestamped client side, so you can deal with out of order messages delivery on the subscriber side
- contain enough information to identify the device. (either by publishing on a topic that holds the device id, or by adding the id in the message's payload)
